File: Day02/part2.py
MIN_ABS_DELTA = 1
MAX_ABS_DELTA = 3

def get_slope(num):
    if num == 0:
        return 0

    return num / abs(num)

def is_valid(delta):
    slope = get_slope(delta)

    if slope == 0:
        return False
    
    min_delta = min(MIN_ABS_DELTA * slope, MAX_ABS_DELTA * slope)
    max_delta = max(MIN_ABS_DELTA * slope, MAX_ABS_DELTA * slope)

    return delta <= max_delta or delta >= min_delta


def main():
    safe_count = 0

    with open('./input.txt') as file:
        for line in file:
            line = line.strip()
            tokens = line.split()
            levels = [int(token) for token in tokens]

            deltas = []
            i = 0
            level_removed = False

            while i < len(levels) - 1:
                current_delta = levels[i + 1] - levels[i]

                # Try removing current and next
                if len(deltas) == 0:
                    deltas.append(current_delta)
                    continue

                prev_delta = deltas[-1]
                prev_slope = get_slope(prev_delta)
                current_slope = get_slope(current_delta)

                if prev_slope != current_slope or not is_valid(current_delta):
                    print("Theres an error", prev_delta, current_delta, is_valid(current_delta))

                deltas.append(current_delta)
                print(deltas)

                i += 1

    print(safe_count)

if __name__ == '__main__':
    main()

File: Day02/part2.pl
use strict;
use warnings;
use List::Util qw(all min max sum any);

my $count = 0;

sub is_safe {
    my @levels = @_;

    my @slopes = map {
        my $delta = $levels[$_ + 1] - $levels[$_];
        $delta == 0 ? 0 : $delta / abs($delta);
    } (0..$#levels - 1);

    my $slope = sum(@slopes) > 0 ? 1 : -1;
    my @limits = (1 * $slope, 3 * $slope);
    my $min_slope = min(@limits);
    my $max_slope = max(@limits);

    return all {
        my $delta = $levels[$_ + 1] - $levels[$_];
        $delta <= $max_slope && $delta >= $min_slope;
    } (0..$#levels - 1);
}

open (my $file, "<", "input.txt") or die $!;

while (my $line = <$file>) {
    my @levels = split(" ", $line);
   
    if (is_safe(@levels)) {
        $count++;
    } else {
        my $safe_with_removal = any {
            my $remove_index = $_;
            my @levels_copy = map {$_ eq $remove_index ? () : $_} (0..$#levels);
            is_safe(@levels_copy);
        } (0..$#levels);

        if ($safe_with_removal) {
            $count++;
        }
    }
}

close ($file);

print("$count\n");

